File: API/PetCenterModels/DataTransferObjects/ProcedureDTO.cs
using Microsoft.VisualBasic;
using PetCenterModels.DBTables;
using PetCenterModels.DataTransferObjects;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Linq;
using System.Text;
using System.Text.Json.Serialization;
using System.Threading.Tasks;
using PetCenterModels.SearchObjects;

namespace PetCenterModels.DataTransferObjects
{
    public class ProcedureSpecificationSubDTO : IBaseResponseDTO<MedicalProcedureSpecification, ProcedureSpecificationSubDTO>
    {
        public Guid? Id {get; set;} = null;
      
        public List<NoteSubDTO>? Notes {get; set;} = null;

        public Guid ProcedureID {get; set;} = Guid.Empty;

        public Guid KindId {get; set;} = Guid.Empty;

        public Guid? BreedId {get; set;} = null;

        public bool Optional {get; set;} = true;

        public bool? SexSpecific {get; set;} = null;

        public int? ApproximateAge {get; set;} = null;

        public short? Interval {get; set;} = null;

        public static ProcedureSpecificationSubDTO? FromEntity(MedicalProcedureSpecification? entity)
        {
            if(entity==null){return null;}
            return new ProcedureSpecificationSubDTO
            {
                Id=entity.Id,
                ProcedureID=entity.ProcedureId,
                Optional=entity.Optional,
                KindId=entity.KindId,
                BreedId=entity.BreedId,
                SexSpecific=entity.SexSpecific,
                Interval=entity.IntervalDays,
                ApproximateAge=entity.ApproximateAge

            };

        }
    }

    public class ProcedureDTO : ISerializableRequestDTO<Procedure>, IBaseResponseDTO<Procedure,ProcedureDTO>
    {
       
        public Guid? Id {get; set;} = null;
      
        public List<NoteSubDTO>? Notes {get; set;} = null;

        public string Description {get; set;} = string.Empty;

        public List<ProcedureSpecificationSubDTO> Specifications {get; set;} = new();

        public static ProcedureDTO? FromEntity(Procedure? entity)
        {
            if(entity==null){return null;}
            ProcedureDTO output = new ProcedureDTO
            {
                Id = entity.Id,
                Description = entity.Description,
                Specifications = entity.Specifications.Select(s=>ProcedureSpecificationSubDTO.FromEntity(s)!).ToList()
            };

            

            return output;
        }

        public Procedure? ToEntity()
        {
            Procedure proc = new();
           
            proc.Description=Description;
            return proc;
        }
        
        
        public bool Validate()
        {
            
            return !string.IsNullOrWhiteSpace(Description);
        }



    }
}

File: Frontend/pet_center_app/lib/models/data_transfer/procedure_dto.dart
import 'package:json_annotation/json_annotation.dart';
import 'package:pet_center_app/models/data_transfer/note_sub_dto.dart';
part 'procedure_dto.g.dart';

@JsonSerializable(explicitToJson: true)
class ProcedureSpecificationSubDTO {
  String? id;
  String currentVersion;
  List<NoteSubDTO>? notes;
  String procedureID;
  String kindId;
  String? breedId;
  bool optional;
  bool? sexSpecific;
  int? approximateAge;
  int? interval;

  ProcedureSpecificationSubDTO({
    this.id,
    this.currentVersion = '',
    this.notes,
    this.procedureID = '',
    this.kindId = '',
    this.breedId,
    this.optional = true,
    this.sexSpecific,
    this.approximateAge,
    this.interval,
  });

  factory ProcedureSpecificationSubDTO.fromJson(Map<String, dynamic> json) =>
      _$ProcedureSpecificationSubDTOFromJson(json);

  Map<String, dynamic> toJson() => _$ProcedureSpecificationSubDTOToJson(this);

  ProcedureSpecificationSubDTO copy() => ProcedureSpecificationSubDTO(
    id: id,
    currentVersion: currentVersion,
    notes: notes?.map((n) => n.copy()).toList(),
    procedureID: procedureID,
    kindId: kindId,
    breedId: breedId,
    optional: optional,
    sexSpecific: sexSpecific,
    approximateAge: approximateAge,
    interval: interval,
  );
}

@JsonSerializable(explicitToJson: true)
class ProcedureDTO {
  String? id;
  String currentVersion;
  List<NoteSubDTO>? notes;
  String description;
  List<ProcedureSpecificationSubDTO> specifications;

  ProcedureDTO({
    this.id,
    this.currentVersion = '',
    this.notes,
    this.description = '',
    List<ProcedureSpecificationSubDTO>? specifications,
  }) : specifications = specifications ?? [];

  factory ProcedureDTO.fromJson(Map<String, dynamic> json) =>
      _$ProcedureDTOFromJson(json);

  Map<String, dynamic> toJson() => _$ProcedureDTOToJson(this);

  ProcedureDTO copy() => ProcedureDTO(
    id: id,
    currentVersion: currentVersion,
    notes: notes?.map((n) => n.copy()).toList(),
    description: description,
    specifications: specifications.map((s) => s.copy()).toList(),
  );
}
